File: main_ort.py
import os
import sys
import time

import cv2
import numpy as np
import onnxruntime as ort

DEFAULT_CUDA_BENCHMARK_ITERATIONS = 20
DEFAULT_CUDA_WARMUP_ITERATIONS = 5
DEFAULT_TIMING_WARMUP_ITERATIONS = 1


def result_path(file_name):
    output_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), "build")
    os.makedirs(output_dir, exist_ok=True)
    return os.path.join(output_dir, file_name)


def load_class_names(model_path):
    names_file_path = os.path.join(os.path.dirname(os.path.abspath(model_path)), "classes.names")
    class_names = {}
    if not os.path.exists(names_file_path):
        print(f"WARN: '{names_file_path}' not found. Using default numeric IDs.")
        return class_names

    with open(names_file_path, "r", encoding="utf-8") as f:
        for idx, line in enumerate(f):
            name = line.strip()
            if name:
                class_names[idx] = name
    return class_names


def hsv_to_rgb(h, s, v):
    rgb_max = v * 2.55
    rgb_min = rgb_max * (100 - s) / 100.0
    i = h // 60
    difs = h % 60
    rgb_adj = (rgb_max - rgb_min) * difs / 60.0

    if i == 0:
        r, g, b = rgb_max, rgb_min + rgb_adj, rgb_min
    elif i == 1:
        r, g, b = rgb_max - rgb_adj, rgb_max, rgb_min
    elif i == 2:
        r, g, b = rgb_min, rgb_max, rgb_min + rgb_adj
    elif i == 3:
        r, g, b = rgb_min, rgb_max - rgb_adj, rgb_max
    elif i == 4:
        r, g, b = rgb_min + rgb_adj, rgb_min, rgb_max
    else:
        r, g, b = rgb_max, rgb_min, rgb_max - rgb_adj

    return int(r), int(g), int(b)


def generate_palette_bgr(size_of_result):
    size = max(1, int(size_of_result))
    palette = []
    for i in range(size):
        h = int(360.0 / size * i)
        r, g, b = hsv_to_rgb(h, 100, 100)
        palette.append((b, g, r))
    return palette


def create_session(model_path, provider):
    available = ort.get_available_providers()
    if provider in ("cuda", "gpu"):
        if "CUDAExecutionProvider" not in available:
            raise RuntimeError(
                "CUDAExecutionProvider is not available. "
                f"onnxruntime={ort.__version__}, available providers={available}"
            )
        providers = ["CUDAExecutionProvider"]
    else:
        providers = ["CPUExecutionProvider"]

    options = ort.SessionOptions()
    options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
    options.intra_op_num_threads = 1
    session = ort.InferenceSession(model_path, sess_options=options, providers=providers)
    if provider in ("cuda", "gpu") and "CUDAExecutionProvider" not in session.get_providers():
        raise RuntimeError(
            "CUDAExecutionProvider could not be activated. "
            f"onnxruntime={ort.__version__}, active providers={session.get_providers()}"
        )
    return session


def run_yolo11_seg_ort_inference(
    model_path,
    image_path,
    show_boxes,
    conf_threshold,
    iou_threshold,
    provider,
    benchmark_iterations=1,
    warmup_iterations=0,
):
    class_names = load_class_names(model_path)
    num_classes = len(class_names) if class_names else 4
    palette_bgr = generate_palette_bgr(num_classes)

    image = cv2.imread(image_path)
    if image is None:
        print(f"ERROR: Could not read image '{image_path}'.")
        return
    original_image = image.copy()
    original_h, original_w = original_image.shape[:2]

    requested_cuda = provider in ("cuda", "gpu")
    benchmark_iterations = max(1, int(benchmark_iterations))
    warmup_iterations = max(0, int(warmup_iterations))

    start_total_time = time.perf_counter()
    session = create_session(model_path, provider)
    print(f" |- Active Providers: {session.get_providers()}")
    input_name = session.get_inputs()[0].name
    output_names = [output.name for output in session.get_outputs()]

    if requested_cuda:
        warmup_blob = cv2.dnn.blobFromImage(original_image, 1 / 255.0, (640, 640), swapRB=True, crop=False)
        try:
            for _ in range(warmup_iterations):
                session.run(output_names, {input_name: warmup_blob})
        except Exception as exc:
            raise RuntimeError(f"CUDA warmup failed: {exc}") from exc

    timing_warmup_blob = cv2.dnn.blobFromImage(original_image, 1 / 255.0, (640, 640), swapRB=True, crop=False)
    try:
        for _ in range(DEFAULT_TIMING_WARMUP_ITERATIONS):
            session.run(output_names, {input_name: timing_warmup_blob})
    except Exception as exc:
        raise RuntimeError(f"Timing warmup failed: {exc}") from exc

    forward_total_ms = 0.0
    pipeline_total_ms = 0.0
    detected_count = 0
    image = original_image.copy()

    for _ in range(benchmark_iterations):
        iter_start = time.perf_counter()
        image = original_image.copy()
        mask_overlay = np.zeros_like(image, dtype=np.uint8)
        blob = cv2.dnn.blobFromImage(image, 1 / 255.0, (640, 640), swapRB=True, crop=False)

        start_forward = time.perf_counter()
        try:
            outputs = session.run(output_names, {input_name: blob})
        except Exception as exc:
            if requested_cuda:
                raise RuntimeError(f"CUDA inference failed: {exc}") from exc
            raise
        end_forward = time.perf_counter()
        forward_total_ms += (end_forward - start_forward) * 1000.0

        if outputs[0].shape[1] == 32:
            proto = outputs[0]
            preds = outputs[1]
        else:
            preds = outputs[0]
            proto = outputs[1]

        preds = np.squeeze(preds)
        if preds.shape[0] < preds.shape[1]:
            preds = preds.T
        proto = np.squeeze(proto)

        boxes = []
        confidences = []
        class_ids = []
        mask_coefficients = []
        x_factor = original_w / 640.0
        y_factor = original_h / 640.0

        for pred in preds:
            scores = pred[4 : 4 + num_classes]
            class_id = int(np.argmax(scores))
            confidence = float(scores[class_id])
            if confidence <= conf_threshold:
                continue

            cx, cy, w, h = pred[0], pred[1], pred[2], pred[3]
            boxes.append([int((cx - w / 2) * x_factor), int((cy - h / 2) * y_factor), int(w * x_factor), int(h * y_factor)])
            confidences.append(confidence)
            class_ids.append(class_id)
            mask_coefficients.append(pred[4 + num_classes :])

        indices = cv2.dnn.NMSBoxes(boxes, confidences, conf_threshold, iou_threshold)
        text_draw_queue = []

        if len(indices) > 0:
            indices = indices.flatten()
            c, proto_h, proto_w = proto.shape
            proto_flat = proto.reshape(c, -1)

            for i in indices:
                left, top, width, height = boxes[i]
                class_id = class_ids[i]
                color = palette_bgr[class_id % len(palette_bgr)]

                coeffs = np.array(mask_coefficients[i])
                single_mask = np.dot(coeffs, proto_flat)
                single_mask = 1 / (1 + np.exp(-single_mask))
                single_mask = single_mask.reshape(proto_h, proto_w)

                scale_mask = cv2.resize(single_mask, (640, 640), interpolation=cv2.INTER_LINEAR)
                mx1, my1 = max(0, int(left / x_factor)), max(0, int(top / y_factor))
                mx2, my2 = min(640, int((left + width) / x_factor)), min(640, int((top + height) / y_factor))

                crop_mask = np.zeros_like(scale_mask)
                crop_mask[my1:my2, mx1:mx2] = scale_mask[my1:my2, mx1:mx2]
                actual_mask = cv2.resize(crop_mask, (original_w, original_h), interpolation=cv2.INTER_LINEAR)
                mask_overlay[actual_mask > 0.5] = color

                if show_boxes:
                    cv2.rectangle(mask_overlay, (left, top), (left + width, top + height), color, 2)

                label = f"{class_names.get(class_id, str(class_id))}: {confidences[i]:.2f}"
                font_scale = 0.55
                (text_w, text_h), baseline = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, font_scale, 1)
                if show_boxes:
                    text_x = left
                    text_y = top - 8
                    if text_y < text_h + 5:
                        text_y = top + height - 8
                        if text_y > original_h - 5:
                            text_y = top + text_h + 5
                else:
                    text_x = left + (width // 2) - (text_w // 2)
                    text_y = top + (height // 2) + (text_h // 2)

                text_x = max(5, min(text_x, original_w - text_w - 5))
                text_y = max(text_h + 5, min(text_y, original_h - baseline - 5))
                text_draw_queue.append((label, (text_x, text_y), font_scale))

            cv2.addWeighted(mask_overlay, 0.4, image, 1.0, 0, image)
            for label, (x, y), scale in text_draw_queue:
                cv2.putText(image, label, (x, y), cv2.FONT_HERSHEY_SIMPLEX, scale, (0, 0, 0), 3, cv2.LINE_AA)
                cv2.putText(image, label, (x, y), cv2.FONT_HERSHEY_SIMPLEX, scale, (255, 255, 255), 1, cv2.LINE_AA)
        else:
            print("WARN: No targets detected.")

        detected_count = len(indices) if len(indices) > 0 else 0
        pipeline_total_ms += (time.perf_counter() - iter_start) * 1000.0

    forward_ms = forward_total_ms / benchmark_iterations
    total_ms = pipeline_total_ms / benchmark_iterations
    print("-------------------------------------------")
    print(f"PERF: Pure inference (forward) cost: {forward_ms:.2f} ms")
    print(f"PERF: Total algorithm pipeline cost: {total_ms:.2f} ms (including pre/post-processing)")
    print(f"INFO: Inference finished. Successfully detected {detected_count} targets.")
    print("-------------------------------------------")

    active_provider = "cuda" if "CUDAExecutionProvider" in session.get_providers() else "cpu"
    output_name = result_path(f"result_onnxruntime_python_{active_provider}.png")
    cv2.imwrite(output_name, image)
    print(f"INFO: Result image saved to: {output_name}")
    if os.environ.get("OPENCV_ONNX_SHOW") == "1":
        try:
            cv2.imshow("ONNX Runtime Python YOLO11-Seg", image)
            cv2.waitKey(0)
            cv2.destroyAllWindows()
        except cv2.error as exc:
            print(f"WARN: Could not open display window: {exc}")


if __name__ == "__main__":
    if len(sys.argv) < 6:
        print("ERROR: Insufficient arguments!")
        print(
            "USAGE: python main_ort.py <model_path.onnx> <image_path.png/.jpg> "
            "<True/False> <conf_threshold> <iou_threshold> [cpu|cuda]"
        )
        sys.exit(-1)

    model_arg = sys.argv[1]
    image_arg = sys.argv[2]
    show_boxes_arg = sys.argv[3].lower() in ["true", "1", "t", "y", "yes"]
    conf_threshold_arg = float(sys.argv[4])
    iou_threshold_arg = float(sys.argv[5])
    provider_arg = sys.argv[6].lower() if len(sys.argv) >= 7 else "cpu"
    benchmark_iterations_arg = DEFAULT_CUDA_BENCHMARK_ITERATIONS if provider_arg in ("cuda", "gpu") else 1
    warmup_iterations_arg = DEFAULT_CUDA_WARMUP_ITERATIONS if provider_arg in ("cuda", "gpu") else 0

    print("INFO: Starting ONNX Runtime Python inference...")
    print(f" |- Model: {model_arg}")
    print(f" |- Image: {image_arg}")
    print(f" |- Provider: {provider_arg}")
    print(f" |- ONNX Runtime Version: {ort.__version__}")
    print(f" |- Available Providers: {ort.get_available_providers()}")
    print(f" |- Show Bounding Boxes: {show_boxes_arg}")
    print(f" |- Confidence Threshold: {conf_threshold_arg}")
    print(f" |- IoU NMS Threshold: {iou_threshold_arg}")
    print(f" |- Benchmark Iterations: {benchmark_iterations_arg}")
    print(f" |- Warmup Iterations: {warmup_iterations_arg}")

    run_yolo11_seg_ort_inference(
        model_arg,
        image_arg,
        show_boxes_arg,
        conf_threshold_arg,
        iou_threshold_arg,
        provider_arg,
        benchmark_iterations_arg,
        warmup_iterations_arg,
    )

File: main_ort.cpp
#include <onnxruntime_cxx_api.h>
#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>

#include <algorithm>
#include <chrono>
#include <cstdlib>
#include <cmath>
#include <fstream>
#include <iostream>
#include <limits.h>
#include <memory>
#include <numeric>
#include <string>
#include <sstream>
#include <sys/stat.h>
#include <unistd.h>
#include <vector>

struct TextObject {
    std::string label;
    cv::Point loc;
    double scale;
};

std::vector<std::string> loadClassNames(const std::string& modelPath) {
    std::vector<std::string> classNames;
    size_t lastSlash = modelPath.find_last_of("/\\");
    std::string baseDir = (lastSlash == std::string::npos) ? "" : modelPath.substr(0, lastSlash + 1);
    std::ifstream ifs(baseDir + "classes.names");
    if (!ifs.is_open()) {
        std::cout << "WARN: '" << baseDir << "classes.names' not found. Using default numeric IDs." << std::endl;
        return classNames;
    }

    std::string line;
    while (std::getline(ifs, line)) {
        if (!line.empty() && line.back() == '\r') {
            line.pop_back();
        }
        if (!line.empty()) {
            classNames.push_back(line);
        }
    }
    return classNames;
}

std::vector<int64_t> getTensorShape(const Ort::Value& tensor) {
    return tensor.GetTensorTypeAndShapeInfo().GetShape();
}

constexpr int kDefaultCudaBenchmarkIterations = 20;
constexpr int kDefaultCudaWarmupIterations = 5;
constexpr int kDefaultTimingWarmupIterations = 1;

void HSVtoRGB(int* r, int* g, int* b, int h, int s, int v) {
    float RGB_min, RGB_max;
    RGB_max = v * 2.55f;
    RGB_min = RGB_max * (100 - s) / 100.0f;

    int i = h / 60;
    int difs = h % 60;
    float RGB_Adj = (RGB_max - RGB_min) * difs / 60.0f;

    switch (i) {
        case 0:
            *r = static_cast<int>(RGB_max);
            *g = static_cast<int>(RGB_min + RGB_Adj);
            *b = static_cast<int>(RGB_min);
            break;
        case 1:
            *r = static_cast<int>(RGB_max - RGB_Adj);
            *g = static_cast<int>(RGB_max);
            *b = static_cast<int>(RGB_min);
            break;
        case 2:
            *r = static_cast<int>(RGB_min);
            *g = static_cast<int>(RGB_max);
            *b = static_cast<int>(RGB_min + RGB_Adj);
            break;
        case 3:
            *r = static_cast<int>(RGB_min);
            *g = static_cast<int>(RGB_max - RGB_Adj);
            *b = static_cast<int>(RGB_max);
            break;
        case 4:
            *r = static_cast<int>(RGB_min + RGB_Adj);
            *g = static_cast<int>(RGB_min);
            *b = static_cast<int>(RGB_max);
            break;
        default:
            *r = static_cast<int>(RGB_max);
            *g = static_cast<int>(RGB_min);
            *b = static_cast<int>(RGB_max - RGB_Adj);
            break;
    }
}

void appendCudaProvider(Ort::SessionOptions& sessionOptions) {
    OrtCUDAProviderOptionsV2* cudaOptions = nullptr;
    const OrtApi& api = Ort::GetApi();
    Ort::ThrowOnError(api.CreateCUDAProviderOptions(&cudaOptions));
    Ort::ThrowOnError(api.SessionOptionsAppendExecutionProvider_CUDA_V2(sessionOptions, cudaOptions));
    api.ReleaseCUDAProviderOptions(cudaOptions);
}

bool hasDisplay() {
    const char* showWindow = std::getenv("OPENCV_ONNX_SHOW");
    return showWindow != nullptr && std::string(showWindow) == "1";
}

std::vector<cv::Scalar> paletteBgr(int sizeOfResult) {
    std::vector<cv::Scalar> colors;
    int size = std::max(1, sizeOfResult);
    colors.reserve(size);
    for (int i = 0; i < size; ++i) {
        int r = 0;
        int g = 0;
        int b = 0;
        int h = static_cast<int>(360.0 / size * i);
        HSVtoRGB(&r, &g, &b, h, 100, 100);
        colors.emplace_back(b, g, r);
    }
    return colors;
}

std::string executableDir() {
    char buffer[PATH_MAX] = {0};
    ssize_t length = readlink("/proc/self/exe", buffer, sizeof(buffer) - 1);
    std::string path = length > 0 ? std::string(buffer, static_cast<size_t>(length)) : std::string(".");
    size_t lastSlash = path.find_last_of("/\\");
    return lastSlash == std::string::npos ? "." : path.substr(0, lastSlash);
}

std::string resultPath(const std::string& fileName) {
    std::string outputDir = executableDir();
    mkdir(outputDir.c_str(), 0755);
    return outputDir + "/" + fileName;
}

int main(int argc, char** argv) {
    if (argc < 6) {
        std::cerr << "ERROR: Insufficient arguments!" << std::endl;
        std::cerr << "USAGE: " << argv[0]
                  << " <model_path.onnx> <image_path.png/.jpg> <true/false> <conf_threshold> <iou_threshold> [cpu|cuda]"
                  << std::endl;
        return -1;
    }

    std::string modelPath = argv[1];
    std::string imagePath = argv[2];
    std::string showBoxesStr = argv[3];
    std::transform(showBoxesStr.begin(), showBoxesStr.end(), showBoxesStr.begin(), ::tolower);
    bool showBoxes = (showBoxesStr == "true" || showBoxesStr == "1" || showBoxesStr == "yes");
    float confThreshold = std::stof(argv[4]);
    float iouThreshold = std::stof(argv[5]);
    std::string provider = argc >= 7 ? argv[6] : "cpu";
    std::transform(provider.begin(), provider.end(), provider.begin(), ::tolower);
    const bool requestedCuda = (provider == "cuda" || provider == "gpu");
    const int benchmarkIterations = requestedCuda ? kDefaultCudaBenchmarkIterations : 1;
    const int warmupIterations = requestedCuda ? kDefaultCudaWarmupIterations : 0;

    std::vector<std::string> classNames = loadClassNames(modelPath);
    int numClasses = classNames.empty() ? 4 : static_cast<int>(classNames.size());

    std::cout << "INFO: Starting ONNX Runtime C++ inference..." << std::endl;
    std::cout << " |- Model: " << modelPath << std::endl;
    std::cout << " |- Image: " << imagePath << std::endl;
    std::cout << " |- Provider: " << provider << std::endl;
    std::cout << " |- Show Bounding Boxes: " << (showBoxes ? "True" : "False") << std::endl;
    std::cout << " |- Confidence Threshold: " << confThreshold << std::endl;
    std::cout << " |- IoU NMS Threshold: " << iouThreshold << std::endl;

    cv::Mat image = cv::imread(imagePath);
    if (image.empty()) {
        std::cerr << "ERROR: Could not read image: " << imagePath << std::endl;
        return -1;
    }
    cv::Mat originalImage = image.clone();
    int originalW = image.cols;
    int originalH = image.rows;

    Ort::Env env(ORT_LOGGING_LEVEL_WARNING, "opencv_onnxruntime");
    bool usingCuda = requestedCuda;
    auto makeSessionOptions = [](bool enableCuda) {
        Ort::SessionOptions options;
        options.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);
        options.SetIntraOpNumThreads(1);
        if (enableCuda) {
            appendCudaProvider(options);
        }
        return options;
    };

    std::unique_ptr<Ort::Session> session;
    try {
        Ort::SessionOptions sessionOptions = makeSessionOptions(requestedCuda);
        session = std::make_unique<Ort::Session>(env, modelPath.c_str(), sessionOptions);
    } catch (const Ort::Exception& e) {
        if (!requestedCuda) {
            std::cerr << "ERROR: Failed to create ONNX Runtime CPU session: " << e.what() << std::endl;
            return -1;
        }
        std::cerr << "ERROR: Failed to create ONNX Runtime CUDA session: " << e.what() << std::endl;
        return -1;
    }
    usingCuda = requestedCuda;
    std::cout << "INFO: Active Provider: " << (usingCuda ? "cuda" : "cpu") << std::endl;

    Ort::AllocatorWithDefaultOptions allocator;
    auto inputNameAllocated = session->GetInputNameAllocated(0, allocator);
    std::string inputName = inputNameAllocated.get();
    size_t outputCount = session->GetOutputCount();
    std::vector<Ort::AllocatedStringPtr> outputNameAllocated;
    std::vector<const char*> outputNames;
    outputNameAllocated.reserve(outputCount);
    outputNames.reserve(outputCount);
    for (size_t i = 0; i < outputCount; ++i) {
        outputNameAllocated.emplace_back(session->GetOutputNameAllocated(i, allocator));
        outputNames.push_back(outputNameAllocated.back().get());
    }

    std::vector<int64_t> inputShape = {1, 3, 640, 640};
    Ort::MemoryInfo memoryInfo = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);

    const char* inputNames[] = {inputName.c_str()};
    if (requestedCuda) {
        cv::Mat warmupBlob;
        cv::dnn::blobFromImage(originalImage, warmupBlob, 1.0 / 255.0, cv::Size(640, 640), cv::Scalar(), true, false);
        size_t warmupTensorSize = static_cast<size_t>(warmupBlob.total());
        Ort::Value warmupTensor = Ort::Value::CreateTensor<float>(
            memoryInfo, reinterpret_cast<float*>(warmupBlob.data), warmupTensorSize, inputShape.data(), inputShape.size());
        for (int i = 0; i < warmupIterations; ++i) {
            session->Run(Ort::RunOptions{nullptr}, inputNames, &warmupTensor, 1, outputNames.data(), outputNames.size());
        }
    }

    cv::Mat timingWarmupBlob;
    cv::dnn::blobFromImage(originalImage, timingWarmupBlob, 1.0 / 255.0, cv::Size(640, 640), cv::Scalar(), true, false);
    size_t timingWarmupTensorSize = static_cast<size_t>(timingWarmupBlob.total());
    Ort::Value timingWarmupTensor = Ort::Value::CreateTensor<float>(
        memoryInfo, reinterpret_cast<float*>(timingWarmupBlob.data), timingWarmupTensorSize, inputShape.data(), inputShape.size());
    for (int i = 0; i < kDefaultTimingWarmupIterations; ++i) {
        session->Run(Ort::RunOptions{nullptr}, inputNames, &timingWarmupTensor, 1, outputNames.data(), outputNames.size());
    }

    std::vector<Ort::Value> outputs;
    std::vector<cv::Rect> boxes;
    std::vector<float> confidences;
    std::vector<int> classIds;
    std::vector<cv::Mat> maskCoefficients;
    std::vector<int64_t> predShape;
    std::vector<int64_t> protoShape;
    double forwardCost = 0.0;
    double totalCost = 0.0;
    int detectedCount = 0;

    for (int iter = 0; iter < std::max(1, benchmarkIterations); ++iter) {
        cv::Mat imageFrame = originalImage.clone();
        cv::Mat maskOverlay = cv::Mat::zeros(imageFrame.size(), imageFrame.type());
        cv::Mat blob;
        cv::dnn::blobFromImage(imageFrame, blob, 1.0 / 255.0, cv::Size(640, 640), cv::Scalar(), true, false);
        size_t inputTensorSize = static_cast<size_t>(blob.total());
        Ort::Value inputTensor = Ort::Value::CreateTensor<float>(
            memoryInfo, reinterpret_cast<float*>(blob.data), inputTensorSize, inputShape.data(), inputShape.size());

        auto totalIterStart = std::chrono::high_resolution_clock::now();
        auto forwardStart = std::chrono::high_resolution_clock::now();
        try {
            outputs = session->Run(
                Ort::RunOptions{nullptr}, inputNames, &inputTensor, 1, outputNames.data(), outputNames.size());
        } catch (const Ort::Exception& e) {
            std::cerr << "ERROR: ONNX Runtime " << (requestedCuda ? "CUDA" : "CPU")
                      << " inference failed: " << e.what() << std::endl;
            return -1;
        }
        auto forwardEnd = std::chrono::high_resolution_clock::now();

        int predIndex = 0;
        int protoIndex = 1;
        std::vector<int64_t> shape0 = getTensorShape(outputs[0]);
        if (shape0.size() >= 2 && shape0[1] == 32) {
            protoIndex = 0;
            predIndex = 1;
        }

        const float* predData = outputs[predIndex].GetTensorData<float>();
        const float* protoData = outputs[protoIndex].GetTensorData<float>();
        predShape = getTensorShape(outputs[predIndex]);
        protoShape = getTensorShape(outputs[protoIndex]);

        int attrs = 0;
        int proposals = 0;
        bool transposed = false;
        if (predShape.size() == 3) {
            if (predShape[1] < predShape[2]) {
                attrs = static_cast<int>(predShape[1]);
                proposals = static_cast<int>(predShape[2]);
                transposed = true;
            } else {
                proposals = static_cast<int>(predShape[1]);
                attrs = static_cast<int>(predShape[2]);
            }
        } else {
            std::cerr << "ERROR: Unsupported prediction output shape." << std::endl;
            return -1;
        }

        boxes.clear();
        confidences.clear();
        classIds.clear();
        maskCoefficients.clear();
        float xFactor = originalW / 640.0f;
        float yFactor = originalH / 640.0f;
        std::vector<float> row(attrs);

        for (int i = 0; i < proposals; ++i) {
            if (transposed) {
                for (int j = 0; j < attrs; ++j) {
                    row[j] = predData[j * proposals + i];
                }
            } else {
                std::copy(predData + i * attrs, predData + (i + 1) * attrs, row.begin());
            }

            cv::Mat scoresMat(1, numClasses, CV_32FC1, row.data() + 4);
            cv::Point classIdPoint;
            double maxClassScore;
            cv::minMaxLoc(scoresMat, nullptr, &maxClassScore, nullptr, &classIdPoint);
            if (maxClassScore <= confThreshold) {
                continue;
            }

            float cx = row[0];
            float cy = row[1];
            float w = row[2];
            float h = row[3];
            boxes.emplace_back(static_cast<int>((cx - w / 2.0f) * xFactor),
                               static_cast<int>((cy - h / 2.0f) * yFactor),
                               static_cast<int>(w * xFactor),
                               static_cast<int>(h * yFactor));
            confidences.push_back(static_cast<float>(maxClassScore));
            classIds.push_back(classIdPoint.x);
            cv::Mat coeffs(1, 32, CV_32FC1, row.data() + 4 + numClasses);
            maskCoefficients.push_back(coeffs.clone());
        }

        std::vector<int> indices;
        cv::dnn::NMSBoxes(boxes, confidences, confThreshold, iouThreshold, indices);
        detectedCount = static_cast<int>(indices.size());

        std::vector<cv::Scalar> colors = paletteBgr(numClasses);

        std::vector<TextObject> textQueue;
        if (!indices.empty()) {
            int protoChannels = static_cast<int>(protoShape[1]);
            int protoH = static_cast<int>(protoShape[2]);
            int protoW = static_cast<int>(protoShape[3]);
            cv::Mat protoFlat(protoChannels, protoH * protoW, CV_32FC1, const_cast<float*>(protoData));

            cv::Mat maskFrame = cv::Mat::zeros(imageFrame.size(), imageFrame.type());

            for (int idx : indices) {
                cv::Rect box = boxes[idx];
                int classId = classIds[idx];
                cv::Scalar color = colors[classId % colors.size()];

                cv::Mat singleMaskMat = maskCoefficients[idx] * protoFlat;
                singleMaskMat = singleMaskMat.reshape(1, protoH);
                for (int r = 0; r < protoH; ++r) {
                    float* ptr = singleMaskMat.ptr<float>(r);
                    for (int c = 0; c < protoW; ++c) {
                        ptr[c] = 1.0f / (1.0f + std::exp(-ptr[c]));
                    }
                }

                cv::Mat scaleMask;
                cv::resize(singleMaskMat, scaleMask, cv::Size(640, 640));
                int mx1 = std::max(0, static_cast<int>(box.x / xFactor));
                int my1 = std::max(0, static_cast<int>(box.y / yFactor));
                int mx2 = std::min(640, static_cast<int>((box.x + box.width) / xFactor));
                int my2 = std::min(640, static_cast<int>((box.y + box.height) / yFactor));

                cv::Mat cropMask = cv::Mat::zeros(640, 640, CV_32FC1);
                if (mx2 > mx1 && my2 > my1) {
                    scaleMask(cv::Rect(mx1, my1, mx2 - mx1, my2 - my1)).copyTo(
                        cropMask(cv::Rect(mx1, my1, mx2 - mx1, my2 - my1)));
                }

                cv::Mat actualMask;
                cv::resize(cropMask, actualMask, imageFrame.size());
                cv::Mat binaryMask = actualMask > 0.5f;
                maskFrame.setTo(color, binaryMask);

                if (showBoxes) {
                    cv::rectangle(maskFrame, box, color, 2);
                }

                std::string nameLabel = (classId < static_cast<int>(classNames.size())) ? classNames[classId] : std::to_string(classId);
                std::string label = nameLabel + ": " + cv::format("%.2f", confidences[idx]);
                double fontScale = 0.55;
                int baseLine;
                cv::Size labelSize = cv::getTextSize(label, cv::FONT_HERSHEY_SIMPLEX, fontScale, 1, &baseLine);
                int textX = showBoxes ? box.x : box.x + (box.width / 2) - (labelSize.width / 2);
                int textY = showBoxes ? box.y - 8 : box.y + (box.height / 2) + (labelSize.height / 2);
                if (showBoxes && textY < labelSize.height + 5) {
                    textY = box.y + box.height - 8;
                    if (textY > originalH - 5) {
                        textY = box.y + labelSize.height + 5;
                    }
                }
                textX = std::max(5, std::min(textX, originalW - labelSize.width - 5));
                textY = std::max(labelSize.height + 5, std::min(textY, originalH - baseLine - 5));
                textQueue.push_back({label, cv::Point(textX, textY), fontScale});
            }

            cv::addWeighted(maskFrame, 0.4f, imageFrame, 1.0f, 0.0, imageFrame);
            for (const auto& textObj : textQueue) {
                cv::putText(imageFrame, textObj.label, textObj.loc, cv::FONT_HERSHEY_SIMPLEX, textObj.scale, cv::Scalar(0, 0, 0), 3, cv::LINE_AA);
                cv::putText(imageFrame, textObj.label, textObj.loc, cv::FONT_HERSHEY_SIMPLEX, textObj.scale, cv::Scalar(255, 255, 255), 1, cv::LINE_AA);
            }

            image = imageFrame;
        }

        auto totalIterEnd = std::chrono::high_resolution_clock::now();
        forwardCost += std::chrono::duration<double, std::milli>(forwardEnd - forwardStart).count();
        totalCost += std::chrono::duration<double, std::milli>(totalIterEnd - totalIterStart).count();
    }

    forwardCost /= std::max(1, benchmarkIterations);
    totalCost /= std::max(1, benchmarkIterations);

    std::cout << "-------------------------------------------" << std::endl;
    std::cout << "PERF: Pure inference (forward) cost: " << cv::format("%.2f", forwardCost) << " ms" << std::endl;
    std::cout << "PERF: Total algorithm pipeline cost: " << cv::format("%.2f", totalCost) << " ms (including pre/post-processing)" << std::endl;
    std::cout << "INFO: Inference finished. Successfully detected " << detectedCount << " targets." << std::endl;
    std::cout << "-------------------------------------------" << std::endl;

    std::string outName = resultPath(usingCuda ? "result_onnxruntime_cpp_cuda.png" : "result_onnxruntime_cpp_cpu.png");
    cv::imwrite(outName, image);
    std::cout << "INFO: Result image saved to: " << outName << std::endl;
    if (hasDisplay()) {
        try {
            cv::imshow("ONNX Runtime C++ YOLO11-Seg", image);
            cv::waitKey(0);
        } catch (const cv::Exception& e) {
            std::cout << "WARN: Could not open display window: " << e.what() << std::endl;
        }
    }
    return 0;
}
